book/src/fastkzg.md

```diff
@@ -0,0 +1,97 @@
+
+Additionally, nearly linear time fast polynomial evaluation and fast polynomial interpolation algorithms allow the dealer to combine all openings belonging to a participant into a single $\mathbb{G}_1$ element, and allow the recipient to fast verify the opening proof.
+
+### Dealer messages
+
+1. The dealer $d$ chooses a uniformly random polynomial $S$ of degree $p$.
+2. KZG commit to $S$ to obtain $\hat{S}$
+3. KZG commit to $S(0)$ to obtain $[S(0)] G_1$
+4. Create an opening proof $\pi_0$ of commitment $\hat{S} - [S(0)] G_1$ opening to $0$ at $0$.
+5. For all $i \in [1,n]$:
+   - create an opening proof $\pi_i$ of commitment $\hat{S}$ at all points $\alpha \in \Omega_i$.
+   - compute $\hat{T_i} = \hat{R} - \hat{S}_i$ where $T_i = R - S_i$
+
+The dealer signs and posts $(\tau,d,\hat{S}, [S(0)] G_1, \pi_0)$ to the blockchain.
+
+For each node $i$, the dealer encrypts (with MAC) to $pk_i$ the message $(\tau, d, \langle S(\alpha) \mid \alpha \in \Omega_i \rangle, \pi_i)$ which consists of:
+
+* $\tau$, the round number
+* $d$, the dealer id
+* $\hat{S}$, the KZG commitment to the share polynomial
+* $\langle S(\alpha) \mid \alpha \in \Omega_i\rangle$, the share evaluations belonging to node $i$.
+* $\pi_{i}$, the opening proof of $\hat{S}$ at $\Omega_i$
+
+For all $i \in [1,n]$ the dealer posts the $i$th message to the blockchain.
+
+The cost of each plaintext is $2$ integers, ?? $\mathbb{F}_r$ elements and $2$ $\mathbb{G}_1$ elements. (TODO)
+
+#### Offline precompute
+
+Much of the dealer's computation can be done offline/precomputed, potentially saving online computation time. The secret polynomial, commitment, and opening proofs can be speculatively computed, subject to maintaining secrecy of the secret data.
+
+### Receiving dealer message
+
+Node $i$ recieves, decrypts, and authenticates $(\tau,d,\hat{S}, \pi_0)$ and $(\tau, d, \hat{S},  \langle s_{\alpha} \rangle, \pi)$ from dealer $d$ through the blockchain.
+
+1. Check that commitment $\hat{S}$ and proof $\pi$ open to $\langle s_{\alpha} \rangle$
+2. Check that commitment $\hat{S} - [S(0)] G_1$ and proof $\pi_0$ opens to $0$ at $0$.
+3. If the decryption or opening check fails, then initiate pre-success or post-success dispute process
+4. Else, the VSS has succeeded for node $i$. Sign and post the ready message $(\tau, d, \hat{S})$.
+
+### Pre-success dispute
+
+In the case where the dealer $d$ posts an invalid distribution of shares, a node can initiate the pre-dispute process as long as no DKG has finalized using that VSS. A node $i$ signs and posts a message $(\text{dispute}, \tau, d, k, \pi)$ where $k$ is the shared ephemeral secret key used to encrypt the message from $d$ to $i$ and $\pi$ is the NIZK proof that it is the ephemeral key. The remaining nodes adjucate the dispute in favor of the dealer or complainer. In case the dealer is found to be faulty, that dealer's VSS is terminated. Additionally, penalties and rewards may be allocated.
```

Comment:
```suggestion
In the case where the dealer \\(d\\) posts an invalid distribution of shares, a node can initiate the pre-dispute process as long as no DKG has finalized using that VSS. A node \\(i\\) signs and posts a message \\((\text{dispute}, \tau, d, k, \pi)\\) where \\(k\\) is the shared ephemeral secret key used to encrypt the message from \\(d\\) to \\(i\\) and \\(\pi\\) is the NIZK proof that it is the ephemeral key. The remaining nodes adjucate the dispute in favor of the dealer or complainer. In case the dealer is found to be faulty, that dealer's VSS is terminated. Additionally, penalties and rewards may be allocated.
```